### PR TITLE
feat: implement atomic vfolder host permission

### PR DESCRIPTION
**Pull request by fregataa**
ref: https://github.com/lablup/backend.ai/issues/753#issuecomment-1260603825

1. Migrate `domains.allowed_vfolder_hosts`, `groups.allowed_vfolder_hosts`, `keypair_resource_policies.allowed_vfolder_hosts` from ARRAY of string to JSONB

etc) 
```
# Before
domains.allowed_vfolder_hosts = ["local:volume1"]
domains.allowed_vfolder_hosts = ["local:volume2"]

# After
domains.allowed_vfolder_hosts = {"local:volume1": ["mount", "create-vfolder"]}
domains.allowed_vfolder_hosts = {"local:volume2": ["read-vfolder", "create-vfolder"]}
```

2. Impl `PermissionListColumn` which can be used by other **permission** spec columns.

### CODE REVIEW

**Comment by kyujin-cho** [comment]
Please set PR status to "Ready for review" when you've done the final touches.

**Comment by fregataa** [comment]
@yomybaby @lizable @Sujin-Kim1 
There should be update in frontend since `allowed_vfolder_hosts` field of `Domain`, `Group`,`KeyPairResourcePolicy` graphene and db schema will change in this PR.
There is an alternative, just letting `allowed_vfolder_hosts` as a legacy field and adding `allowed_vfolder_host_map` JSON field.

Please give me your opinion.

**Review by achimnol — COMMENTED** [review]
We need to add an API to retrieve the list of possible permission names for the web UI and control panel, to avoid duplicated constant definitions.

File: src/ai/backend/manager/models/acl.py
Comment (kyujin-cho) [inline src/ai/backend/manager/models/acl.py]: `Permission` only represents capability of vFolder host feature. Please rename it relevantly to a new name which represents only permission of vFolder hosts.

File: src/ai/backend/manager/models/acl.py
Comment (kyujin-cho) [inline src/ai/backend/manager/models/acl.py]: Plural suggestion!
```suggestion
def get_all_permissions() -> Mapping[str, Any]:
```

File: src/ai/backend/manager/models/acl.py
Comment (kyujin-cho) [inline src/ai/backend/manager/models/acl.py]: This should be vFolder "host", not just vFolder.
```suggestion
        "vfolder_host_permission_list": [perm.value for perm in VFolderHostPermission],
```

**Review by kyujin-cho — CHANGES_REQUESTED** [review]

File: src/ai/backend/manager/models/acl.py
Comment (fregataa) [reply src/ai/backend/manager/models/acl.py]: Ok. I am going to rename it. I intended to add more permissions in this **general** graphene object when we implement ACL in the future.

File: src/ai/backend/client/cli/admin/acl.py
Comment (kyujin-cho) [inline src/ai/backend/client/cli/admin/acl.py]: As this command is already a subset of `admin` group I think it's okay to remove this message.
```suggestion
```

File: src/ai/backend/client/cli/admin/domain.py
Comment (kyujin-cho) [inline src/ai/backend/client/cli/admin/domain.py:65]: We need to tell user how to fill out this parameter. After reading actual implementation it looks like the code expects string formatted as JSON syntax to be filled here, so let's tell user that way.

File: src/ai/backend/client/cli/admin/domain.py
Comment (kyujin-cho) [inline src/ai/backend/client/cli/admin/domain.py]: Same here.

File: src/ai/backend/client/cli/admin/group.py
Comment (kyujin-cho) [inline src/ai/backend/client/cli/admin/group.py:91]: Also here.

File: src/ai/backend/client/cli/admin/group.py
Comment (kyujin-cho) [inline src/ai/backend/client/cli/admin/group.py]: Same here.

File: src/ai/backend/client/cli/admin/resource_policy.py
Comment (kyujin-cho) [inline src/ai/backend/client/cli/admin/resource_policy.py]: Hmm, shouldn't it be `volume1:local`?

**Review by kyujin-cho — CHANGES_REQUESTED** [review]

File: src/ai/backend/client/cli/admin/resource_policy.py
Comment (fregataa) [reply src/ai/backend/client/cli/admin/resource_policy.py]: but the original default value was `["local"]`. Is it good to change?

Comment (kyujin-cho) [reply src/ai/backend/client/cli/admin/resource_policy.py:101]: Yes. The value added on database has format of `host:volume`, so the example should follow that rule also.

File: src/ai/backend/client/func/acl.py
Comment (lizable) [inline src/ai/backend/client/func/acl.py]: According to the query schema defined [here](https://github.com/lablup/backend.ai/pull/790/files#diff-56a2d9f1fc07827156313028f4414a518eb3f14bc56739d16d6bbb3e32956019R490-R493), the query name should be changed to "vfolder_host_permission", not "permissions".

File: src/ai/backend/client/func/acl.py
Comment (lizable) [inline src/ai/backend/client/func/acl.py]: Of course, the return type should be changed, too.
`permissions` to `vfolder_host_permissions`.

**Review by lizable — CHANGES_REQUESTED** [review]
I found that there are a few updates on vfolder host permissions.
It's such a small update, but crucial for those who need to use that API.

File: src/ai/backend/client/func/keypair_resource_policy.py
Comment (lizable) [inline src/ai/backend/client/func/keypair_resource_policy.py:58]: Btw, as long as I know,  both`total_resource_slots` and `allowed_vfolder_hosts` have the same type (stringified JSON). Therefore, the type definition for `total_resource_slots` needs to be changed likewise.

File: src/ai/backend/client/func/keypair_resource_policy.py
Comment (lizable) [inline src/ai/backend/client/func/keypair_resource_policy.py:106]: Here as well. The type definition of `total_resource_slots` is `str`, not `int`.

**Review by lizable — CHANGES_REQUESTED** [review]
IMO, there are a few requests written in the comments and also one more thing to be fixed before merging this PR.
TL;DR: We need to change the code in here(`src/ai/backend/manager/model/gql.py#1428`)
https://github.com/lablup/backend.ai/blob/40c04807d9b03040218b39ead81262ff483e79f2/src/ai/backend/manager/models/gql.py#L1415
to this
`mutation_cls = getattr(Mutations, info.field_name).type`

Considering the overall key `allowed_vfolder_hosts` are located in several policies (such as `domains`, `groups(project groups`, and `keypair_resource_policies`), I think It's quite crucial to show the overview of access in those policies easily.
In light of this, It's not surprising that users may want features for updating those access controls at once. Therefore, I decided to chain mutations so that It reduces(not completely, but some) repetitive requests to manager server. To make chained mutations, it needs [aliasing](https://atheros.ai/blog/how-to-use-graphql-aliases). Unfortunately, In current code, It regards aliasing string as one of the variables in Mutations class, which would definitely throw an error. I checked many graphql requests and It doesn't affect any mutations at all.
It would be appreciated you to apply the new code described above.

**Comment by lizable** [comment]
> IMO, there are a few requests written in the comments and also one more thing to be fixed before merging this PR. TL;DR: We need to change the code in here(`src/ai/backend/manager/model/gql.py#1428`)
> 
> https://github.com/lablup/backend.ai/blob/40c04807d9b03040218b39ead81262ff483e79f2/src/ai/backend/manager/models/gql.py#L1415
> 
> 
> to this
> `mutation_cls = getattr(Mutations, info.field_name).type`
> Considering the overall key `allowed_vfolder_hosts` are located in several policies (such as `domains`, `groups(project groups`, and `keypair_resource_policies`), I think It's quite crucial to show the overview of access in those policies easily. In light of this, It's not surprising that users may want features for updating those access controls at once. Therefore, I decided to chain mutations so that It reduces(not completely, but some) repetitive requests to manager server. To make chained mutations, it needs [aliasing](https://atheros.ai/blog/how-to-use-graphql-aliases). Unfortunately, In current code, It regards aliasing string as one of the variables in Mutations class, which would definitely throw an error. I checked many graphql requests and It doesn't affect any mutations at all. It would be appreciated you to apply the new code described above.

It's better to handle this in as a separate issue(#879), which is already opened by @yomybaby .

File: src/ai/backend/client/func/keypair_resource_policy.py
Comment (lizable) [reply src/ai/backend/client/func/keypair_resource_policy.py:58]: @fregataa FWIW, shouldn't be `total_resource_slots` and `allowed_vfolder_hosts` to be the same type?

Comment (lizable) [reply src/ai/backend/client/func/keypair_resource_policy.py:58]: You already applied it [here](https://github.com/lablup/backend.ai/pull/790/files#diff-4f594e20f517e33a816d296b20b711339b333fdfbbaf42dfa95c648682cac272R104).

**Review by lizable — CHANGES_REQUESTED** [review]
@fregataa 
There's a minor request about correcting type definition on the parameter.
Except that, It works fine.

File: src/ai/backend/client/func/keypair_resource_policy.py
Comment (fregataa) [reply src/ai/backend/client/func/keypair_resource_policy.py:58]: I read how keypair_resource_policy mutation works and it seems `total_resource_slots` should not be `None` because the field determine the policy. but I guess `total_resource_slots` can be `None` so I leave it Optional.

Plus, group and domain are different from keypair_resource_policy. they can have `total_resource_slots = None`

Comment (lizable) [reply src/ai/backend/client/func/keypair_resource_policy.py:58]: I get it now. It should be left as is.

**Review by lizable — APPROVED** [review]
LGTM.
Btw I dunno why a couple of irrelevant test codes failed, but I guess we can ignore this.
And also, I guess we need one more approval from reviewers (probably @kyujin-cho or @achimnol  or @adrysn ) since this change is epic.

**Review by kyujin-cho — APPROVED** [review]

**Pull request merged**
